compiler/rustc_lint/src/non_local_def.rs

```diff
@@ -0,0 +1,89 @@
+use rustc_hir::{Body, Item, ItemKind, QPath, TyKind};
+use rustc_span::{sym, MacroKind};
+
+use crate::{lints::NonLocalDefinitionsDiag, LateContext, LateLintPass, LintContext};
+
+declare_lint! {
+    /// The `non_local_definitions` lint checks for `impl` blocks and `#[macro_export]`
+    /// macro inside bodies (functions, enum discriminant, ...).
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// trait MyTrait {}
+    /// struct MyStruct;
+    ///
+    /// fn foo() {
+    ///     impl MyTrait for MyStruct {}
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Creating non local definitions go against expectation and can create decrepencies
+    /// in tooling. In should be avoided.
+    pub NON_LOCAL_DEFINITIONS,
+    Warn,
+    "checks for non local definitions"
+}
+
+#[derive(Default)]
+pub struct NonLocalDefinitions {
+    is_in_body: u32,
+}
+
+impl_lint_pass!(NonLocalDefinitions => [NON_LOCAL_DEFINITIONS]);
+
+impl<'tcx> LateLintPass<'tcx> for NonLocalDefinitions {
+    fn check_body(&mut self, _cx: &LateContext<'tcx>, _body: &'tcx Body<'tcx>) {
+        self.is_in_body += 1;
+    }
+
+    fn check_body_post(&mut self, _cx: &LateContext<'tcx>, _body: &'tcx Body<'tcx>) {
+        self.is_in_body -= 1;
+    }
+
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'tcx>) {
+        if self.is_in_body > 0 {
+            match item.kind {
+                ItemKind::Impl(impl_) => 'block: {
+                    let TyKind::Path(QPath::Resolved(_, self_ty_path)) = impl_.self_ty.kind else {
+                        break 'block;
+                    };
```

Comment:
I don't think this is enough. Here are some cases that I think should still lint:

```rust
fn meow() {
  struct Cat;
  impl From<Cat> for () { /* .. */ }
}

fn woof() {
  struct Dog;
  impl PartialEq<Dog> for &Dog { /* .. */ }
}

struct Wrap<T>(T);

fn rawr() {
  struct Lion;
  impl From<Wrap<Wrap<Lion>>> for { /* .. */ }
}
```

Reply:
Also, the RFC says
> unless either the Trait or the Type is also nested inside the same expression-containing item.

so
```rust
struct CuteCat;
fn uwu() {
  trait YarnBall {}
  impl YarnBall for CuteCat {}
}
```
should _not_ lint.

Reply:
> should _not_ lint.

and it doesn't. See the test file:

```rust
struct Test;

// ...

const _: () = {
    trait Uto {}

    impl Uto for Test {}
};
```

Reply:
> ```rust
> fn meow() {
>   struct Cat;
>   impl From<Cat> for () { /* .. */ }
> }
> 
> fn woof() {
>   struct Dog;
>   impl PartialEq<Dog> for &Dog { /* .. */ }
> }
> 
> struct Wrap<T>(T);
> 
> fn rawr() {
>   struct Lion;
>   impl From<Wrap<Wrap<Lion>>> for { /* .. */ }
> }
> ```

I don't think any of those 3 tests should fire since the `impl` blocks are at the same level as the structs. You can't write:

```rust
fn meow() {
   struct Cat;
}

impl From<Cat> for () { /* .. */ }
```

you are obligated to put it at least at the same level, but you shouldn't put in a nested block. So this should (and does) fire the lint:

```rust
fn meow() {
   struct Cat;

   fn bar() {
       impl From<Cat> for () { /* .. */ }
   }
}
```

Reply:
I messed up my examples, I did mean to put the structs outside the function. I don't see you looking at the substs in the lint, so I don't think that fires? (without testing it locally)

Reply:
Oh no! I had completely forgotten about generic args in path.

So I went to fix it and spent 2 and a half hours trying to fix it with iterator combinators, lots of `map`, `chain`, `flatten`, ... it was beautiful! Until I realised that I had again forgotten "about generic args in path": it's recursive, ah!!!! 😩

So now I have to write a recursive function, I take the time to clean up the code, try hard to finally realise that I wanted the "Logical NOR" (deuh 😲).

Anyway, fixed! 😌